src/schoology-extractor/schoology_extractor/schoology_request_service.py

```diff
@@ -0,0 +1,112 @@
+# SPDX-License-Identifier: Apache-2.0
+# Licensed to the Ed-Fi Alliance under one or more agreements.
+# The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+# See the LICENSE and NOTICES files in the project root for more information.
+
+from dataclasses import dataclass
+from logging import Logger
+from typing import Any, Dict, List
+
+import pandas as pd
+
+from .api.request_client import RequestClient
+from .mapping import users as usersMap
+
+
+@dataclass
+class SchoologyRequestService:
+    logger: Logger
+    request_client: RequestClient
+    page_size: int
+
+    @property
+    def _logger(self):
+        assert isinstance(self.logger, Logger)
+        return self.logger
+
+    @property
+    def _client(self):
+        assert isinstance(self.request_client, RequestClient)
+        return self.request_client
+
+    @property
+    def _page_size(self):
+        assert isinstance(self.page_size, int)
+        return self.page_size
+
+    def get_users(self) -> pd.DataFrame:
+
+        self._logger.debug("Exporting users: get users")
+        users_response = self._client.get_users(self._page_size)
+        users_list: List[Any] = []
+        while True:
+            users_list = users_list + users_response.current_page_items
+            if users_response.get_next_page() is None:
+                break
+
+        self._logger.debug("Exporting users: get roles")
+        roles_list: List[Any] = []
+        roles_response = self._client.get_roles(self._page_size)
+        while True:
+            roles_list = roles_list + roles_response.current_page_items
+            if roles_response.get_next_page() is None:
+                break
+
+        users_df = pd.DataFrame(users_list)
+        roles_df = pd.DataFrame(roles_list)
+
+        return usersMap.map_to_udm(users_df, roles_df)
+
+    def get_sections(self) -> list:
+
+        self._logger.debug("Exporting sections: get active courses")
+        courses_response = self._client.get_courses(self._page_size)
+        courses_list: List[Any] = []
+        while True:
+            courses_list = courses_list + courses_response.current_page_items
+            if courses_response.get_next_page() is None:
+                break
+
+        self._logger.debug("Exporting sections: get sections for active courses")
+
+        # TODO: inconsistency - in this case the pagination is handled inside of
+        # the client, instead of being handled here. Arguably there should be
+        # some other thing, neither this service nor the client, that handles
+        # the paging. As BB said in a pr comment, perhaps it should be handled
+        # in a decorator. This works but we should consider refactoring to a
+        # cleaner approach.
+        return self._client.get_section_by_course_ids([c["id"] for c in courses_list])

```

Comment:
@lalrg  @bradbanister please note this TODO comment. I am going to create a stand-alone ticket to handle refactoring these existing methods. Luis, with your work next week you'll be adding new resources that need paging. The decorator idea is interesting but that might be a lot for any of us to take in this new language, especially without anyone else to talk through it with.

May I suggest that the paging for now should go into the `RequestClient`? I can see arguments either way. Ultimately, I think this domain service layer should be insulated from details about paging, and therefore the paging support needs to be pushed into the API client.

Reply:
https://tracker.ed-fi.org/browse/FIZZ-142